release-automations/create_release_branch.py

```diff
@@ -0,0 +1,527 @@
+#!/usr/bin/env python3
+"""
+This script automates the creation of release branches for theRock projects and their submodules.
+It supports branching from a specified source branch, including or excluding specific projects,
+and optionally notifying stakeholders via email.
+
+Main Features:
+- Clones or uses an existing source tree for theRock and its submodules.
+- Checks out the specified source branch for all relevant projects.
+- Determines projects to branch based on .gitmodules and user input.
+- Creates a plan for branching, including checks for branch existence and push permissions.
+- Displays a summary of planned actions
+- Executes branch creation and pushes to remote repositories.
+- Sends notifications to specified email addresses.
+
+Usage:
+    python create_release_branch.py --release_branch <branch> [options]
+
+Options:
+    --release_branch         Name of the branch to create (required).
+    --apitoken              API token for authentication with upstream repositories.
+    --staging               Use staging environment.
+    --prod                  Use production environment.
+    --mailing_list          List of email addresses to notify.
+
+Note:
+    - This script assumes a specific repository structure and the presence of .gitmodules.
+"""
+import argparse
+import configparser
+import glob
+import logging
+import os
+import shutil
+import subprocess
+import tempfile
+import pandas as pd
+import smtplib
+from email.mime.text import MIMEText
+from email.mime.multipart import MIMEMultipart
+
+logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')
+
+class rockbrachingAutomation():
+
+    def __init__(self, args):
+        
+        self.release_branch = args.release_branch
+        if args.apitoken:
+            self.apitoken = args.apitoken
+        else:
+            self.apitoken = None
+        self.is_staging = bool(args.staging)
+        self.is_prod = bool(args.prod)
+        
+        if args.staging:
+            self.gh_hostname = "github.amd.com"
+            self.org = "ROCm-Staging"
+        if args.prod:
+            self.gh_hostname = "github.com"
+            self.org = "ROCm"
+        # Derive environment label for notifications
+        if self.is_staging:
+            self.environment = "Staging"
+        elif self.is_prod:
+            self.environment = "Prod"
+        else:
+            self.environment = "Unknown"
+        
+        if args.prod and args.staging:
+            logging.error("You cannot pass both --staging(-S) and --prod(-P); choose one")
+            exit(1)
```

Comment:
Prefer `parser.error()` during argparse code for preconditions like this. See https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.error

Code example: https://github.com/ROCm/TheRock/blob/1814579823cdb6aecd682eda3478cd2667106c6d/build_tools/compute_rocm_package_version.py#L153-L160